### PR TITLE
Add description of the provider suspension process

### PR DESCRIPTION
**Pull request by potiuk**
Following discussion at the devlist, we are adding description of the suspension process for providers that hold us back from upgrading old dependencies. Discussion here:

https://lists.apache.org/thread/j98bgw9jo7xr4fvjh27d6bfoyxr1omcm

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: README.md
Comment (dstandish) [inline README.md]:
```suggestion
The suspension may be triggered by any committer after the following criteria are met:
```

File: README.md
Comment (vincbeck) [inline README.md:503]: I would increase the resolving period to something more than 1 week. To me, it feels a bit short (especially if the provider does not get much traction) to resolve the dependencies. I dont have any specific period in mind but something like 2 weeks or 1 month seems more reasonable to me. WDYT?

File: README.md
Comment (dstandish) [inline README.md]:
```suggestion
* Explanation why we need to suspend the provider is stated in a public discussion in the devlist. Followed
```

File: README.md
Comment (dstandish) [inline README.md]:
```suggestion
  by LAZY CONSENSUS or VOTE (with the majority of the voters agreeing that we should suspend the provider)
```

i think majority of committers means you require at least half of all commiters to vote in favor!

File: README.md
Comment (dstandish) [inline README.md:509]: do we need to clarify whether any vote is required or what the process is for lifting?

File: README.md
Comment (dstandish) [reply README.md:503]: my guess is that you don't want it to block release for too long, and since it's reversible it's not a "big deal"....

File: README.md
Comment (ferruzzi) [reply README.md]: Revert that last change, s/provide/provider :P

File: README.md
Comment (potiuk) [reply README.md:503]: That was indeed my point (no big deal). I think (and I also generally did not describe it) there are few types of response:

1) resolving it (cool - we do nothing) 
2) response "we are working it, coming soon" in which case we may 
        * decide to wait if we trust things will get resolved
        * or suspend it anyway (because we think it will take more time or we don't trust the promise
3)  no response at all

I think it is really case-by-case and in case of some combinations of those responses/reputation of a dependency/complexity of the change we might decide to wait way longer than 1 week or be absolutely sure that if we don't get response in 1 week, fix will never happen.

For example I have a very strong feeling we will not hear back anything in a week for https://forums.mysql.com/read.php?50,708413) - whereas we already got some feedback here in yandex https://github.com/yandex-cloud/python-sdk/issues/71 and we already know that apache-beam change is already merged and waits for release https://github.com/apache/airflow/pull/30067  and we know that google team works on updating all libraries in https://github.com/apache/airflow/pull/30067. All of those require a bit different approach/timing. I am quite sure we won't susspend google provider off even if the target for the fix is May. But we might decide to suspend mysql 1 week after notice (old mysql provider will work and it has nothing to do with mysql support by airflow as metadata). 

This is BTW a nice spectrum of projects/responses :)  - all of them are about protobuf version requirement. 

That's why I put "at least 1 week". To leave it open-ended but to put some bare minimum waiting time. Sometimes we might simply want to unblock things fast.

File: README.md
Comment (potiuk) [reply README.md:509]: I think we do not need to vote for it, this is basically a normal process of approving/rejecting PR I **think** - I think it should be - in general - way easier to come back if you fulfill the criteria of not holding us back and you already "were" a member. What happens for example if dependency is fixed and the majority of voters will say "no" ? I think it is a bit unfair if we call it "suspension" where it might be a "removal" - and the "come back" should be guaranteed when the conditions are fulfilled.

I think we might want to define a process of removal of providers (I hinted that in the description) and then indeed, bringing the providers back should require a vote as with a new provider (which I think we should start doing) but I think that is not that day to define it (yet).

Comment (potiuk) [reply README.md:509]: But I am open for other suggestions too :)

File: README.md
Comment (hussein-awala) [reply README.md:503]: I wonder if we can find a method to early detect the dependencies which could block the providers release (for ex two major versions older than the latest version), in this case we can create warn issues and invite the maintainers of dependencies of the provider to update its dependencies.

File: README.md
Comment (hussein-awala) [inline README.md:490]: > in the future we might add process to remove them

IMHO this could complicate the process to solve the dependencies problem, except if we wait for a reasonable period (ex 2-3 minor versions for Airflow) without any reactivity from the dependencies maintainers.

File: README.md
Comment (hussein-awala) [reply README.md:509]: I agree with @potiuk
>  I think it is a bit unfair if we call it "suspension" where it might be a "removal" - and the "come back" should be guaranteed when the conditions are fulfilled.

Also, it may have side effects on other suspended providers, where maintainers will be discouraged to avoid wasting time fixing dependencies at the risk of not taking their changes. IMO we need to be clear on this, the suspension will be lifted once the providers dependencies are up to date.

File: README.md
Comment (r-richmond) [inline README.md]: Curious on the reasoning for excluding patch releases `#.#.1`  i.e. With the current example it means google-provider can't update until 2.6.0 which leaves the cases like the [google-ads](https://github.com/apache/airflow/pull/30067#issuecomment-1488521445) open for that much longer.

File: README.md
Comment (potiuk) [reply README.md:490]: > I wonder if we can find a method to early detect the dependencies which could block the providers release (for ex two major versions older than the latest version), in this case we can create warn issues and invite the maintainers of dependencies of the provider to update its dependencies.

Yes this is already partially happening (and I think it's "good enough". We could potentially try to do it automatically to improve the current process. 

BUT

This is a complex and difficult (and I think impossible) matter to do it automatically - it took `pip` two years and a [grant](https://pyfound.blogspot.com/2019/12/moss-czi-support-pip.html) of ~ 400.000 USD from Mozilla Foundation and Chan-Zuckerberg initiative to hash out algorithm that tries to find and resolve any conflicts (and even that was not enough the effort including fixing a lot of teething problems took far more than that. I think proactively detecting and flagging keeping and flagging such problems two major version (or anything like that) before is probably at least an order of magnitude more complex task. And no-one of us would will to do it "proactively" and manually check 650+ dependencies we have.

But I think it is also something we already do manually when we find those. Generally speaking we **know** (at least people who are contributing to providers) that some dependencies are holding us back for some cases. Good example was a google provider. In October 2022 I created the issue about google provider dependencies (related to our Python 3.11 support) https://github.com/apache/airflow/issues/27292 and I pinged google team several times since then stating that those have to be solved, as this is not only blockin 3.11 but also cause the "ads" problem to materialize (as it did). Our system is designed around automating and testing dependencies when they **can** be automatically upgraded and tested, and I think sometimes we have good reasons to stay behind, I think we will be able to detect and flag cases that need attention quite early and create.

So yeah agree we **should** give early warnings about those and in most cases we do. And again - the "effect" of suspension on it's own is nothing more than a very strong warning. It does not have immediate effects - it's just excluding the provider from the future release and blocking any new changes coming. And a clear sign to maintainers "you have to do something".

Comment (potiuk) [reply README.md:490]: > > in the future we might add process to remove them

> IMHO this could complicate the process to solve the dependencies problem, except if we wait for a reasonable period (ex 2-3 minor versions for Airflow) without any reactivity from the dependencies maintainers.

Yes. That's why we are not defining it (yet). 

I think removal of provider from our code should be much more deliberate and serious decisions will have to be made on simply removing the provider and moving it to "attic" (this is the term that the Apache Software Foundation uses) https://attic.apache.org/. And we are nowhere near (I think) to be close to a decision of doing so for any of our providers.

As it works in the foundation Each PMC can release multiple "artifacts". Each "arifact" (Source package) can be moved separately to the Attic when the PMC decides in voting. This is an established process in the foundation to move the whole PMC to attic (https://attic.apache.org/process.html) - but I discussed it with the "elders" in the Foundation and we could move any of our artifacts to the Attic and follow the same principles. 

But - as I stated above - I think we are nowhere near to remove any of our artifacts to the Attic. I think if it happens that we will have a suspended provider for 2-3 months for example and we ping the maintainers of related dependencies and we don't find another solution to "resume" the provider, we might want to move it to the attic (but I think rather than deciding on the policy upfront, we might define the policy as we see it is needed. This is the usual approach we have here, we are aimging to be (both Foundation and airflow) rather light on policies and we rather wokr here in the way that we define the policy the first or even second time we have to deal with an issue (unless this is something we can foresee will impact our users if there is no clear policy).

Comment (potiuk) [reply README.md:490]: > Also, it may have side effects on other suspended providers, where maintainers will be discouraged to avoid wasting time fixing dependencies at the risk of not taking their changes. IMO we need to be clear on this, the suspension will be lifted once the providers dependencies are up to date.

Yep. I think it is quite clear as it is defined now:

> The suspension will be lifted when the dependencies of the provider are made compatible with the Apache
> Airflow and with other providers.

There is no other conditions and the suspension "will" be lifted "when" -> fixing dependencies is the only condition here.

File: README.md
Comment (potiuk) [reply README.md]: @r-richmond. You are right. This is a good point. In this case we might also decide to make the suspension effective in the patch release.

Just to be clear - we are NOT going to suspend "google" provider. That's out of the question and Google is committed to upgrade all dependencies :). In fact the whole process here is to make it possible for them to upgrade, because currently it's `yandex` and `mysql` and `apache-beam` (on top of the dependency upgrades that google team does are holding back `google` provider from being upgraded. And the direction is clear:

The `yandex`, `mysql` and `beam` are restricting dependencies (protobuf <4 generally speaking) that make it impossible for `google` to upgrade their dependencies (generally to upgrade all clients that require protobuf >=4). So `google' provider need is the driving force for considering suspension of those three providers.

And strictly speaking, there is nothing "technology" side that would prevent us from doing so. Technically speaking it will happen automatically after we cherry-pick the "suspension" change to the 2-N-test branch. And we already agreed before and documented that **just** dependency in constraints (even breaking) change is not a breaking nor even significant  change for airflow and it happens all the time in the patchlevel releases. Note that users will still be able to (even if the constraints will be different) in most cases to install older provider versions together - for example even if new google provider introduces protobuf >=4, then users will **still** be able to install previous google provider + latest (suspended) yandes + latest (suspended) mysql : if we decide to suspend them. They just won't be able to upgrade to latest google if they decide to stay with yandex/mysql.

So yeah. I am in for making it even default to remove it from the next patchlevel release.

**Comment by potiuk** [comment]
Following the comment of @r-richmond I updated the process to reflect that we might want to include suspension in the patchlevel releases as well.

File: README.md
Comment (uranusjr) [reply README.md:509]: It may be a good idea to add some text guiding potential contributors what to do? Adding a line saying _If you wish to see a provider to exit suspention, please consider the following steps xxx yyy_.

Comment (potiuk) [reply README.md:509]: Yes. Good point @uranusjr . I will describe the technical details/instrunctions in the PRs that will follow when I implement all the changes needed in our provider release process .

File: README.md
Comment (vincbeck) [reply README.md:503]: Makes sense @potiuk. I was not sure if the one week period included the fix itself or just a ping back to see the next steps. I am totally fine with your reasoning and agree with it :)

File: README.md
Comment (o-nikolas) [inline README.md:491]: How are we going to enforce this one?

Comment (potiuk) [reply README.md:491]: I will add pre-commit to fail such PR.

Comment (potiuk) [reply README.md:491]: (and it will indicate to the reviewer not to merge it).

**Comment by potiuk** [comment]
The LAZY CONSENSUS is reached - looking for approvals here.

**Review by vincbeck — APPROVED** [review]

**Review by ferruzzi — APPROVED** [review]
I read it over and followed the discussion, any questions I had were already covered and I just realized I never actually hit that +1(non-binding)

**Comment by potiuk** [comment]
Are we good to go ?

**Review by hussein-awala — APPROVED** [review]

**Review by uranusjr — APPROVED** [review]

**Review by josh-fell — APPROVED** [review]

**Pull request merged**
